### ISSUE TITLE
[circle-quantizer] Support GPTQ

### ISSUE THREAD

**Issue by 01000-you** [Samsung/ONE#13480]
**What**
- We propose supporting the GPTQ algorithm, a state-of-the-art post-training quantization (PTQ) method that has demonstrated robust performance, 
effectively compressing weights. Notably, GPTQ shows significant efficacy with quantization levels down to 4 bits and even 3 bits on a group-wise basis. 

  Paper: https://arxiv.org/abs/2210.17323

**Why**
- The upcoming AI models support weight formats of up to 4 bits. To align with these changes, 
we need to integrate high-performance quantization algorithms such as GPTQ.

**How**
- The implementation will involve converting the Python code provided by the paper's authors into C++, 
and integrating it with the existing ONE codebase. The current processes in record-minmax and QuantDequant
weightPass offer a procedural foundation suitable for the GPTQ algorithm which compute the Hessian via input and update weights to compensate for quantization errors.

**Method**
- The Hessian matrix is computed by averaging the values from 2X*X^T for every 100 input samples.
- Weight updating and quantization are executed row by row with compensation applied to each row 
to address quantization errors. This adjustment is carried out on a layer-by-layer basis.

**Overview**
    ![_gptq](https://github.com/user-attachments/assets/05e894c2-a620-49fb-bbf1-2bfacf863bdf)

**Comment by 01000-you** [comment]
### Experimental results
- The models below are vision classification models provided by [PyTorch](https://pytorch.org/vision/stable/models.html#classification).
- We evaluated only using weight quantization.

  | Size |   |   | Accuracy |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | --
  | NumParams | FP32 size | INT4 size | Baseline | PTQ-W8 | **GPTQ-W8** | PTQ-W4 | **GPTQ-W4**
DeiT | 5,000,000 | 19.07MB | 4.77MB | 0.7202 | 0.7201 | 0.7202 | 0.6466 | 0.6918
EfficientFormer | 12,290,000 | 46.88MB | 11.72MB | 0.8018 | 0.8002 | 0.8017 | 0.2023 | 0.77
ResNet18 | 11,689,512 | 44.59MB | 11.15MB | 0.6976 | 0.6974 | 0.6973 | 0.5821 | 0.6879
ResNet50 | 25,557,032 | 97.49MB | 24.37MB | 0.7615 | 0.7607 | 0.7611 | 0.5821 | 0.7557
RegNet400mf | 4,344,144 | 16.57MB | 4.14MB | 0.7403 | 0.7395 | 0.7404 | 0.3613 | 0.7194
ResNeXt50 | 25,028,904 | 95.48MB | 23.87MB | 0.7761 | 0.7758 | 0.7763 | 0.6559 | 0.7686
Wide ResNet50 | 68,883,240 | 262.77MB | 65.69MB | 0.7848 | 0.7849 | 0.7847 | 0.7114 | 0.7801
Vgg16 | 138,357,544 | 527.79MB | 131.95MB | 0.7159 | 0.7156 | 0.7158 | 0.4644 | 0.6992
SqueezeNet | 1,248,424 | 4.76MB | 1.19MB | 0.581 | 0.5796 | 0.5803 | 0.3335 | 0.5609
ShuffleNet_x0_5 | 1,366,792 | 5.21MB | 1.30MB | 0.6055 | 0.6021 | 0.6043 | 0.1033 | 0.3634

**Comment by jinevening** [comment]
The result in https://github.com/Samsung/ONE/issues/13480#issuecomment-2270215801 shows that GPTQ is effective in 4 bit weight quantization. For 8bits, the current PTQ works well for all benchmark models.

Do you have a plan to support 4 bit weight quantization?

**Comment by seanshpark** [comment]
> Do you have a plan to support 4 bit weight quantization?

@01000-you , this was asked several months ago. it would help if you provide some information.

**Comment by seanshpark** [comment]
@01000-you , @lemmaa , @jinevening and I had a short talk about this task.

What is your future plan with `record-hessian` tool and adding this feature in `circle-quantizer` ?
- can we quantize all the Ops?
- or is it for particular model?
- what is your plan for Ops that are not supported as of now?

**Comment by seanshpark** [comment]
Can https://github.com/Samsung/ONE/issues/13480#issuecomment-2270215801 experiment results can be reproduced with draft #13585 ?

**Comment by 01000-you** [comment]
@jinevening 
I apologize for the delayed response. As you mentioned, there is not much benefit for 8-bit quantization. 
However, we have considered supporting 4-bit quantization in the future. Recently, weight quantization is challenging even lower than 4-bit quantization. 
while most models show significant performance degradation when quantized to 4-bit, 
GPTQ significantly improves this issue. GPTQ algorithm is almost de facto used in the domain of weight quantization. 
Therefore, we propose that it can be optionally used when 4-bit quantization is supported in the future, 
while currently working with 8-bit as default.

**Comment by 01000-you** [comment]
@seanshpark 
GPTQ can apply convolutions and FC layers only. As now we only support regular Conv2d and FC layer for any models.
The op sets not supported as of now, will covered by in the same way as the circle quantizer.
We haven't experimented with the models for VD customers, but if you suggest one we will run experiments.

**Comment by seanshpark** [comment]
> The op sets not supported as of now, will covered by in the same way as the circle quantizer.

OK. I'll understand GPTQ will quantize for Conv2D and FC and other nodes as existing quantization flows.

Why does `compiler/luci/pass/src/QuantizeWeightsWithGPTQPass.cpp` file process other Ops?

**Comment by seanshpark** [comment]
> What we did was fake quantization similar to how QuantizeDeQuantizeWeightPass works.

So there is no 4bit quantized model?

**Comment by seanshpark** [comment]
> You can reproduce it using circle-quantizer --quantize_dequantize_weights_with_gptq float32 uint4 channel --config ...

I'm not good at quantization. plz provide full description.

**Comment by seanshpark** [comment]
> I will send 4bit model via email.

You don't need to send to email. I don't want to make addition channels for discussion.

```
circle-quantizer \
--quantize_dequantize_weights_with_gptq float32 uint4 channel \
<input_model_path> <output_model_path> \
--input_data <input_data_path>
```

I'd like to try with well known models with current draft #13585.

1. please rebase the draft to current head(master branch)
2. I'd like to try with Inception v3 model. I'll add the results here.
3. there is `--input_data <input_data_path>` . how do I produce this data file for IV3?
4. How do I know GPTQ is working correctly?

**Comment by 01000-you** [comment]
### Guide to GPTQ Quantization for Inception V3 (WIP)

[val.zip](https://github.com/user-attachments/files/17903829/val.zip)

**Comment by 01000-you** [comment]
@seanshpark 

I just posted an example on how to try GPTQ with Inception_v3. Due to the 25MB upload limit, I had to reduce the size of the input data. Please check it out and let me know.

**Comment by seanshpark** [comment]
@01000-you , can you plz rebase #13585 to current master?

**Comment by 01000-you** [comment]
@seanshpark, 
Yes, I just did it. Please note that the removal of the ```cal_offset_2d``` function has not yet been reflected.

**Comment by seanshpark** [comment]
you have rebased to recent master, but just rebased and you didn't check is compiles or not.
this makes me impossible to check the results of your tasks that makes me frustrated.
you let me notify you even with basic things, that makes me very stressed for several months.
I'm closing this issue as I assume you don't have further willing to continue on this task.
if you really do have, you can reopen and continue.
